### ISSUE TITLE
Understanding LOFO Importance

### ISSUE THREAD

**Issue by r0f1** [aerdem4/lofo-importance#17]
Hi,
Consider the following code:

```python
import pandas as pd
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import KFold
from lofo import LOFOImportance

brca = load_breast_cancer()
df = pd.DataFrame(brca["data"], columns=brca["feature_names"]).assign(target=brca["target"])

lofo = LOFOImportance(df, brca["feature_names"], "target", 
                      cv=KFold(n_splits=5, random_state=4), 
                      scoring="roc_auc")
importance_df = lofo.get_importance()
```
[image1](https://imgur.com/a/DSo9y3b)

This produces a importance ranking of features. There are some positive means and some negative means in the dataset. 

I will now select all features, whose mean is above zero and calculate the importance again.

```python
new_features = importance_df.query("importance_mean > 0")["feature"].tolist()
lofo = LOFOImportance(df, new_features, "target", 
                      cv=KFold(n_splits=5, random_state=4), 
                      scoring="roc_auc")
importance_df2 = lofo.get_importance()
```

[image2](https://imgur.com/a/3tAP8CE)

Again, there are some positive means and some negative means in the dataset. This is not what I was expecting. I removed the supposedly unnecessary features after the first step, why are there new unnecessary features after the second step?

**Comment by aerdem4** [comment]
Hi @r0f1, thanks for letting us know. As far as I see, the feature importances are very low and their standard deviations are relatively high. This means that positive ones are not significantly positive and negative ones are not significantly negative. Can you also make the same test by providing a model you know (instead of default model) which has high AUC on this dataset?

**Comment by r0f1** [comment]
Hi,
So you would recommend not to drop features if the standard deviation is too high? 

I ran a couple of experiments on the BRCA dataset. I found that the default RandomForestClassifier has a good accuracy (0.912) and roc_auc (0.989).  Running the LOFO importance with this model gave me unexpected results: Almost all features are in the red. 

In order to do my feature selection procedure (=take only those features whose mean is above zero, irgnoring standard deviation), I switched to accuracy. Running the LOFO importance with accuracy, over halt the features were green, and others were red. I removed the red ones and re-ran the LOFO imporance on the green ones only, and then there were again some red ones. 

Also, between the runs, the ordering of the features changed.

For fun, I calculated the permutation importance also, and the orderings a completely different from the LOFO importance orderings. Are these the results of random variability? Is there maybe another dataset that you would recommend to re-run those experiments?

Code to reproduce everything: 
[repo](https://github.com/r0f1/lofotest), [binder](https://mybinder.org/v2/gh/r0f1/lofotest/1b16f3dd3b1aee8fa5b6af151241adc1eb212cdc
)

**Comment by aerdem4** [comment]
Thanks for the extensive explanation. I have just checked the dataset. Almost all the features are over 0.9 correlated with another feature. Therefore removing one feature out doesn't have any significant effect. This is the reason why LOFO has small values as output with relatively high standard deviations. So while reading LOFO's output, please always consider the standard deviations.

Some dataset examples:
https://www.kaggle.com/divrikwicky/pf-f-lofo-importance-on-adversarial-validation
https://www.kaggle.com/divrikwicky/lofo-importance
https://www.kaggle.com/divrikwicky/santanderctp-lofo-feature-importance

**Comment by r0f1** [comment]
Aha ok. So I learned two things: Checking for correlated features before applying LOFO and not to ignore the standard deviation in the LOFO result dataframe :) Also, thank you for the additional links. I guess this is closed now. Thanks.
